### PR TITLE
ARROW-11675: [CI][C++] Resolve ctest failures on VS 2019 builds

### PR DESCRIPTION
**Pull request by ianmcook**

### CODE REVIEW

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by pitrou** [comment]
Wow, the build took two hours. Is there a way to make it shorter?

**Comment by pitrou** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by pitrou** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by pitrou** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by ianmcook** [comment]
@pitrou the manifest file `cpp/vcpkg.json` specifies which Arrow library build dependencies to install when `ARROW_DEPENDENCY_SOURCE=VCPKG`, as described in the [Building Arrow C++](https://arrow.apache.org/docs/developers/cpp/building.html) and [Developing on Windows](https://arrow.apache.org/docs/developers/cpp/windows.html) docs, so if you're making changes to that file, please keep in mind that it serves a general purpose beyond what this CI job tests.

**Comment by pitrou** [comment]
Yawn. When I added some logging calls inside the inner routines, the bug disappeared. Looks like a compiler bug of sorts :-(

**Comment by pitrou** [comment]
TODO: perhaps add Spaced tests and specialization

**Comment by pitrou** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by pitrou** [comment]
@github-actions crossbow submit test-build-vcpkg-win

File: dev/tasks/vcpkg-tests/github.windows.yml [hunk 1]
```diff
@@ -36,6 +36,7 @@ jobs:
           git -C arrow checkout FETCH_HEAD
           git -C arrow submodule update --init --recursive
       - name: Remove and Reinstall vcpkg
+        # TODO: this comment seems outdated, the vcpkg on GHA is quite recent.
```
Comment (pitrou) [inline dev/tasks/vcpkg-tests/github.windows.yml:39]: Note @kou @kszucs : I tried to remove the vcpkg-reinstall step, but then got an error:
```
Error: Couldn't find explicitly specified baseline `"a267ab118c09f56f3dae96c9a4b3410820ad2f0b"` in the baseline file, and there was no baseline at that commit or the commit didn't exist.
Error: while checking out baseline 'a267ab118c09f56f3dae96c9a4b3410820ad2f0b:versions/baseline.json':
fatal: path 'versions/baseline.json' exists on disk, but not in 'a267ab118c09f56f3dae96c9a4b3410820ad2f0b'
```
https://github.com/ursacomputing/crossbow/runs/2933445520?check_suite_focus=true

**Review by bkietz — APPROVED** [review]
A few nits. Otherwise this looks fine. Thanks for figuring this out

**Comment by ianmcook** [comment]
Thanks for working on this! The boost changes in `cpp/vcpkg.json` look good 👍

File: dev/tasks/vcpkg-tests/github.windows.yml (hunk 1 above)
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/github.windows.yml:39]: @pitrou versions of vcpkg rapidly stop working until updated, and the most reliable way to update vcpkg is simply to remove and reinstall it, as is done here. I'm not surprised that removing the reinstall step would trigger an error.

The comment lines 39–45 could be removed. I think the final two lines of this comment (`# When running...`) are probably sufficient.

File: cpp/src/parquet/statistics.cc [hunk 2]
```diff
@@ -83,12 +87,25 @@ struct UnsignedCompareHelperBase {
...
+  // implementation-defined if the original value does not fix in the signed type
+  // (i.e., two's complement cannot be assumed even on mainstream machines,
+  // because the compiler may decide otherwise).  Hence the use of `SafeCopy`
+  // below for deterministic bit-casting.
+  // (see "Integer conversions" in
+  //  https://en.cppreference.com/w/cpp/language/implicit_conversion)
+
+  static const T DefaultMin() { return SafeCopy<T>(std::numeric_limits<UCType>::max()); }
+  static const T DefaultMax() {
+    return SafeCopy<T>(std::numeric_limits<UCType>::lowest());
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/parquet/statistics.cc:103]: Entirely right. I'll do.

**Comment by pitrou** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by pitrou** [comment]
CI failures are unrelated, I'll merge.

**Pull request closed without merging**